presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java

```diff
@@ -1479,6 +1485,77 @@ protected Node aggregateResult(Node aggregate, Node nextResult)
+            check(!escapeString.isEmpty(), "Empty Unicode escape character", context);
+            check(escapeString.length() == 1, "Invalid Unicode escape character: " + escapeString, context);
+            escape = escapeString.charAt(0);
+            check(isValidUnicodeEscape(escape), "Invalid Unicode escape character: " + escapeString, context);
+        }
+        else {
+            escape = '\\';
+        }
+
+        String rawContent = unquote(context.UNICODE_STRING().getText().substring(2));
+        StringBuilder unicodeStringBuilder = new StringBuilder();
+        StringBuilder escapedCharacterBuilder = new StringBuilder();
+        int charactersNeeded = 0;
+        boolean isCharacterEscaped = false;
+        for (int i = 0; i < rawContent.length(); i++) {
+            char ch = rawContent.charAt(i);
+            if (ch == escape) {
+                check(escapedCharacterBuilder.length() == 0, "Incomplete escaped sequence: " + escapedCharacterBuilder.toString(), context);
+                if (isCharacterEscaped) {
+                    unicodeStringBuilder.append(escape);
+                }
+                isCharacterEscaped = !isCharacterEscaped;
+                continue;
+            }
+
+            if (!isCharacterEscaped) {
+                unicodeStringBuilder.append(ch);
+                continue;
+            }
+
+            if (charactersNeeded == 0) {
+                if (ch == '+') {
+                    charactersNeeded = 6;
+                    continue;
+                }
+                else {
+                    charactersNeeded = 4;
+                }
+            }
+            check(isHexDigit(ch), "Invalid hexadecimal digit: " + ch, context);
+            escapedCharacterBuilder.append(ch);
+
+            if (escapedCharacterBuilder.length() != charactersNeeded) {
+                continue;
+            }
+
+            String currentEscapedCode = escapedCharacterBuilder.toString();
+            escapedCharacterBuilder.setLength(0);
+
+            int codePoint = Integer.parseInt(currentEscapedCode, 16);
+            check(Character.isValidCodePoint(codePoint), "Invalid escaped character: " + currentEscapedCode, context);
+            if (Character.isSupplementaryCodePoint(codePoint)) {
+                unicodeStringBuilder.appendCodePoint(codePoint);
+            }
+            else {
+                char currentCodePoint = (char) codePoint;
+                check(!Character.isSurrogate(currentCodePoint), format("Invalid escaped character: %s. Escaped character is a surrogate. Use '\\+123456' instead.", currentEscapedCode), context);
+                unicodeStringBuilder.append(currentCodePoint);
+            }
+            charactersNeeded = 0;
+            isCharacterEscaped = false;
+        }
+        check(escapedCharacterBuilder.length() == 0 && !isCharacterEscaped, "Incomplete escaped sequence: " + escapedCharacterBuilder.toString(), context);
+        return unicodeStringBuilder.toString();
```

Comment:
I'm glad that you demonstrated desire to clean up this code.

But I regret to say that this is still not a good shape. Primarily, there are two confusion points:
* You used a potential of 2 or 3 variables to represent your state. This makes it hard for the reader to reason.
* The order of your code is very delicate. What goes first, what corner cases need to be checked, when (not) to fallthrough.

I took the liberty to restructure this. I'm sorry. I mean no offense.

```
        UnicodeDecodeState state = EMPTY;
        for (int i = 0; i < rawContent.length(); i++) {
            char ch = rawContent.charAt(i);
            switch (state) {
                case EMPTY:
                    if (ch == escape) {
                        state = ESCAPED;
                    }
                    else {
                        unicodeStringBuilder.append(ch);
                    }
                    break;
                case ESCAPED:
                    if (ch == escape) {
                        unicodeStringBuilder.append(escape);
                        state = EMPTY;
                    }
                    else if (ch == '+') {
                        state = UNICODE_SEQUENCE;
                        charactersNeeded = 6;
                    }
                    else if (isHexDigit(ch)) {
                        state = UNICODE_SEQUENCE;
                        charactersNeeded = 4;
                        escapedCharacterBuilder.append(ch);
                    }
                    else {
                        throw parseError("Invalid hexadecimal digit: " + ch, context);
                    }
                    break;
                case UNICODE_SEQUENCE:
                    check(isHexDigit(ch), "Incomplete escape sequence: " + escapedCharacterBuilder.toString(), context);
                    escapedCharacterBuilder.append(ch);
                    if (charactersNeeded == escapedCharacterBuilder.length()) {
                        String currentEscapedCode = escapedCharacterBuilder.toString();
                        escapedCharacterBuilder.setLength(0);
                        int codePoint = Integer.parseInt(currentEscapedCode, 16);
                        check(Character.isValidCodePoint(codePoint), "Invalid escaped character: " + currentEscapedCode, context);
                        if (Character.isSupplementaryCodePoint(codePoint)) {
                            unicodeStringBuilder.appendCodePoint(codePoint);
                        }
                        else {
                            char currentCodePoint = (char) codePoint;
                            check(!Character.isSurrogate(currentCodePoint), format("Invalid escaped character: %s. Escaped character is a surrogate. Use '\\+123456' instead.", currentEscapedCode), context);
                            unicodeStringBuilder.append(currentCodePoint);
                        }
                        state = EMPTY;
                        charactersNeeded = -1;
                    }
                    break;
                default:
                    throw new UnsupportedOperationException();
            }
        }

        check(state == UnicodeDecodeState.EMPTY, "Incomplete escape sequence: " + escapedCharacterBuilder.toString(), context);